<a href="https://colab.research.google.com/github/madhavadama/tensorflow/blob/master/Session_7_Classification_MNIST_CNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MNIST Classification using CNN

Step 1 : Collect Data

In [0]:
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets('MNIST_data', one_hot=True)

Instructions for updating:
Please use alternatives such as official/mnist/dataset.py from tensorflow/models.
Instructions for updating:
Please write your own downloading logic.
Instructions for updating:
Please use urllib or similar directly.
Successfully downloaded train-images-idx3-ubyte.gz 9912422 bytes.
Instructions for updating:
Please use tf.data to implement this functionality.
Extracting MNIST_data/train-images-idx3-ubyte.gz
Successfully downloaded train-labels-idx1-ubyte.gz 28881 bytes.
Instructions for updating:
Please use tf.data to implement this functionality.
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Instructions for updating:
Please use tf.one_hot on tensors.
Successfully downloaded t10k-images-idx3-ubyte.gz 1648877 bytes.
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Successfully downloaded t10k-labels-idx1-ubyte.gz 4542 bytes.
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz
Instructions for updating:
Please use alternatives such as official/mnist/dataset.py fr

Step 1a : Get Training and Test Data

In [0]:
trainX = mnist.train.images
trainY = mnist.train.labels

In [0]:
testX = mnist.test.images
testY = mnist.test.labels

Set the Seed

In [0]:
import numpy as np

In [0]:
np.random.seed(42)

In [0]:
trainX.shape

(55000, 784)

# Build the Graph

import Keras Models and Layers

In [0]:
from tensorflow.python.keras.models import Sequential

In [0]:
from tensorflow.python.keras.layers import Dense, Dropout, Conv2D, Flatten, Reshape, MaxPool2D

Start a Sequential model

In [0]:
model = Sequential()

Reshape Data

In [0]:
# first level flattening , for it be proccessed in Conv Layer 
model.add(Reshape((28,28,1),input_shape=(784,)))

Add first Convolutional layer : 32 filters 3x3 with stride as 1x1

In [0]:
# if you have color image, then use Conv3D 
model.add(Conv2D(32,kernel_size=(5,5),activation='relu',strides=(1,1)))
#26x26x32

Add 2nd Convolutional layer : 64 filters 3x3 with stride as 1x1

In [0]:
model.add(Conv2D(64,kernel_size=(5,5),strides=(1,1),activation='relu'))
#24x24x64

Add MaxPooling layer

In [0]:
model.add(MaxPool2D(pool_size=(2,2)))
#6x6x64

Add Dropout

In [0]:
# model.add(Dropout(0.25))

Flatten the Output

In [0]:
model.add(Flatten())

Add Fully Connected Layer - 128 neurons

In [0]:
# fully connected layer 
model.add(Dense(128,activation='relu'))

Another Dropout

In [0]:
# model.add(Dropout(0.2))
#Suppose you have 128 units in the layer and set the keep_prob to 0.5, Then the activation of 64 randomly chosen units out of 128 will be set to 0, and the remaining 64 will be scaled by a factor of 50. I think a more precise description is that you only keep the activation of 50 percent of the nodes.

Add Softmax Layer

In [0]:
model.add(Dense(10,activation='softmax'))

Compile model    

In [0]:
model.compile(optimizer='adam',loss='categorical_crossentropy',metrics=['accuracy'])

Train

In [0]:
model.fit(trainX,trainY,epochs=20,batch_size=500,
          validation_data=(testX,testY))

Train on 55000 samples, validate on 10000 samples
Epoch 1/20
55000/55000 [==============================] - 4s 81us/sample - loss: 0.3117 - acc: 0.9093 - val_loss: 0.0769 - val_acc: 0.9767
Epoch 2/20
55000/55000 [==============================] - 4s 77us/sample - loss: 0.0643 - acc: 0.9812 - val_loss: 0.0397 - val_acc: 0.9875
Epoch 3/20
55000/55000 [==============================] - 4s 77us/sample - loss: 0.0459 - acc: 0.9864 - val_loss: 0.0358 - val_acc: 0.9879
Epoch 4/20
55000/55000 [==============================] - 4s 77us/sample - loss: 0.0332 - acc: 0.9901 - val_loss: 0.0299 - val_acc: 0.9894
Epoch 5/20
55000/55000 [==============================] - 4s 77us/sample - loss: 0.0265 - acc: 0.9921 - val_loss: 0.0284 - val_acc: 0.9896
Epoch 6/20
55000/55000 [==============================] - 4s 77us/sample - loss: 0.0207 - acc: 0.9935 - val_loss: 0.0309 - val_acc: 0.9894
Epoch 7/20
55000/55000 [==============================] - 4s 77us/sample - loss: 0.0174 - acc: 0.9945 - val_loss: 0.

3 x 3 Filer 



55000/55000 [==============================] - 5s 82us/sample - loss: 0.0309 - acc: 0.9897 - val_loss: 0.0321 - val_acc: 0.9894
<tensorflow.python.keras.callbacks.History at 0x7ff16cf6f470>



5 x 5 


Epoch 20/20
55000/55000 [==============================] - 5s 94us/sample - loss: 0.0111 - acc: 0.9962 - val_loss: 0.0197 - val_acc: 0.9936
<tensorflow.python.keras.callbacks.History at 0x7ff120d1b518>



Total params: 229,706
Trainable params: 229,706
Non-trainable params: 0




# 5 x 5 without droput 


Epoch 20/20
55000/55000 [==============================] - 4s 77us/sample - loss: 0.0025 - acc: 0.9991 - val_loss: 0.0327 - val_acc: 0.9915
<tensorflow.python.keras.callbacks.History at 0x7ff1207d2208>


Total params: 872,714
Trainable params: 872,714
Non-trainable params: 0



Saving Model as HDF5 file

In [0]:
model.save('keras_cnn.h5')

Show Model

In [0]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
reshape_2 (Reshape)          (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 24, 24, 32)        832       
_________________________________________________________________
conv2d_7 (Conv2D)            (None, 20, 20, 64)        51264     
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 10, 10, 64)        0         
_________________________________________________________________
flatten_2 (Flatten)          (None, 6400)              0         
_________________________________________________________________
dense_4 (Dense)              (None, 128)               819328    
_________________________________________________________________
dense_5 (Dense)              (None, 10)               